### ISSUE TITLE
Callback URL ignores NEXTAUTH_URL

### ISSUE THREAD

**Issue by balazsorban44** [nextauthjs/next-auth#587]
**Describe the bug**
Updated from 2.x to 3.x and changed references from `site` to use the `NEXTAUTH_URL` env variable.
When trying to log in on a deployed app, I can verify that the `NEXTAUTH_URL` is set correctly (same as the base URL of the deployment, eg.: `https://example.com`), but when going to the url `https://example.com/api/auth/signin/identity-server4`, I am redirected to `localhost:3000`.

The `https://example.com/api/auth/session` request has a cookie header `next-auth.callback-url=http%3A%2F%2Flocalhost%3A3000` as well.

**Screenshots or error logs**
An image of `https://example.com/api/auth/providers`
![image](https://user-images.githubusercontent.com/18369201/90500796-abd6cf80-e14b-11ea-824a-d69d1601656d.png)

**UPDATE:**
Downgrading to `2.2.0` with the following diff makes it work again, so I am almost certain it is an actual bug in `next-auth`:
(`APP_HOST` equals to what `NEXTAUTH_URL` should be, so `https://example.com`)
`package.json`
![image](https://user-images.githubusercontent.com/18369201/90616695-0a648200-e20e-11ea-8379-a0d9d89b4bae.png)

`pages/api/auth/[...nextauth.ts]`
![image](https://user-images.githubusercontent.com/18369201/90616462-bfe30580-e20d-11ea-94ea-3cf7b31b1ec6.png)

`pages/_app.tsx`
![image](https://user-images.githubusercontent.com/18369201/90616553-dab57a00-e20d-11ea-818b-10a26c817756.png)

**Additional context**
Apart from removing the `site` references and adding `NEXTAUTH_URL`, nothing has been changed.

**Feedback**
*Documentation refers to searching through [online documentation](https://next-auth.js.org), code comments and issue history. The example project refers to [next-auth-example](https://github.com/iaincollins/next-auth-example).*

* [x] Found the documentation helpful
* [ ] Found documentation but was incomplete
* [ ] Could not find relevant documentation
* [ ] Found the example project helpful
* [x] Did not find the example project helpful

I am wondering if someone else has had a similar issue while upgrading from 2.x to 3.x, and if there are any tests confirming that the issue is on my side, or maybe there are some old references somewhere that overrides the `NEXTAUTH_URL`?

**Comment by iaincollins** [comment]
You say the example project was not helpful, are you able to replicate this issue with the example project?

**Comment by balazsorban44** [comment]
I will see if I can deploy the example project, the only thing that looks different for me is that I deploy to Azure, not to Vercel. But I can 100% verify that the `NEXTAUTH_URL` is set correctly in the deployed app environment. (I have an `/api/env` route, which logs certain env variables to a logging service, nothing to the user, just for debugging purposes.)

**Comment by MehediH** [comment]
I can confirm I am facing the same issue. My production site is deployed on Vercel, and it seems to ignore the `NEXTAUTH_URL` environment variable. Redirects to localhost when trying to log in on the production site.

**Comment by iaincollins** [comment]
@MehediH The example site is deployed on Vercel. This is working in the current release and can be verified on https://next-auth-example.now.sh 

I'm going to close this as there is no way to replicate the reported and we can confirm we don't see any issues with this with the example project, which is both a live demo and an open source project, for complete transparency.

As we can show it working (with source) we need a link to source of a repo that replicates the issue you are seeing, ideally also an instance of the site with a problem that we can refer to, otherwise we are only able to guess at what the problem might be, as it's something we don't see in the example project / live demo.

In the case of Azure, there may well be issues with build process which would cause a problem, but we'd need more info about the process to help (e.g. is it a Docker container? Is it Serverless? how is it packaged?); specifically we'd need access to the repository and/or access to an example that shows the problem in action.

**Comment by MehediH** [comment]
@iaincollins, thank you for pointing me to the example repo. Could you please share how you have set the environment variables on the Vercel dashboard? I have the `NEXTAUTH_URL` variable set under the Environment Variables section (for both Prod and Preview) like below:

![image](https://user-images.githubusercontent.com/10274434/92228117-3d11aa00-ee9f-11ea-8894-1007a573207a.png)

Also, here is the repo in question: https://github.com/MehediH/winstall/tree/vnext

And here is the auth config: https://github.com/MehediH/winstall/blob/vnext/pages/api/auth/%5B...nextauth%5D.js

I'll look into this further, but if you can take a quick look to see if I'm doing anything wrong, I'd much appreciate it!

**Comment by MehediH** [comment]
@iaincollins thank you! Looks like I had to set the env variable for both production and dev environments in the Vercel dashboard.

**Comment by balazsorban44** [comment]
@iaincollins Just to ellaborate here, unfortunatelly I was not able to create a reproduction, so I totally understand the decision for closing this issue. Although, after some trial and error, I found the following:

My `next.config.js` included this line:
```js
module.exports = {
  env: {
     NEXTAUTH_URL: process.env.NEXTAUTH_URL
  }
}
```

This way, I could set my env variables in my build pipeline variables (in Azure DevOps), and did not need to set them in Azure App Services' configuration. Apart from `NEXTAUTH_URL`, I set around 20 variables the same way, without an issue. Heck, even `NEXTAUTH_URL` looked fine, as I created an endpoint `/api/env` that logs some of my variables to a logging service, and `NEXTAUTH_URL` showed the correct url.

But for some reason, I actually had to set the `NEXTAUTH_URL` in Azure App Service for it to be correctly detected!

I cannot explain the reason, as said everything else works, but I am planning to move my variables to Azure App Service configurations anyway, so this issue will be resolved, I just wanted to collaborate, maybe someone else has a better explanation.
